private/buf/buflsp/buflsp_test.go

```diff
@@ -0,0 +1,278 @@
+				"objc_class_prefix",
+				"edition_defaults",
+				"packed",
+				"retention",
+				"idempotency_level",
+				"php_namespace",
+				"php_metadata_namespace",
+				"weak",
+				"features",
+				"py_generic_services",
+				"no_standard_descriptor_accessor",
+				"targets",
+				"java_package",
+			},
+		},
+	} {
+		testCase := testCase
+		t.Run(testCase.prefix, func(t *testing.T) {
+			t.Parallel()
+			lspServer.lock.Lock()
+			defer lspServer.lock.Unlock()
+			expectCompletions(t, lspServer, entry, testCase.prefix, testCase.expected)
+		})
+	}
+}
+
+func expectCompletions(t *testing.T, lspServer *server, entry *fileEntry, prefix string, expectedParts []string) {
+	t.Helper()
+	completions := lspServer.findPrefixCompletions(context.Background(), entry, symbolName{"buf", "lsp", "test", "v1"}, prefix)
+	for _, expectedPart := range expectedParts {
+		if _, ok := completions[expectedPart]; !ok {
+			got := make([]string, 0, len(completions))
+			for key := range completions {
+				got = append(got, key)
+			}
+			t.Fatalf("expected %q in completions, got %v", expectedPart, got)
+		}
+		delete(completions, expectedPart)
+	}
+	if len(completions) != 0 {
+		got := make([]string, 0, len(completions))
+		for key := range completions {
+			got = append(got, key)
+		}
+		t.Fatalf("got unexpected completions: %v", got)
+	}
+}
+
+func newTestBufLspWith(t *testing.T, fileName string) (*server, protocol.DocumentURI, error) {
+	t.Helper()
+	lspServer, err := newTestBufLsp(t)
+	if err != nil {
+		return nil, "", err
+	}
+	entry, err := openFile(context.Background(), lspServer, fileName)
+	if err != nil {
+		return nil, "", err
+	}
+	return lspServer, entry, nil
+}
+
+func openFile(ctx context.Context, lspServer *server, fileName string) (protocol.DocumentURI, error) {
+	fileReader, err := os.Open(fileName)
+	if err != nil {
+		return "", err
+	}
+
+	fileData, err := io.ReadAll(fileReader)
+	if err != nil {
+		return "", err
+	}
+
+	absPath, err := filepath.Abs(fileName)
+	if err != nil {
+		return "", err
+	}
+	fileURI := protocol.DocumentURI("file://" + absPath)
+	if err := lspServer.DidOpen(ctx, &protocol.DidOpenTextDocumentParams{
+		TextDocument: protocol.TextDocumentItem{
+			URI:  fileURI,
+			Text: string(fileData),
+		},
+	}); err != nil {
+		return "", err
+	}
+	return fileURI, nil
+}
+
+func newTestBufLsp(tb testing.TB) (*server, error) {
+	tb.Helper()
+	use := "test"
+	stdout := bytes.NewBuffer(nil)
+	stderr := bytes.NewBuffer(nil)
+	env := newEnvFunc(tb, "")(use)
+
+	appContainer := app.NewContainer(
+		env,
+		nil,
+		stdout,
+		stderr,
+		"test",
+	)
+
+	logger, err := zaputil.NewLoggerForFlagValues(appContainer.Stderr(), "info", "text")
+	if err != nil {
+		return nil, err
+	}
+	verbosePrinter := verbose.NewPrinter(appContainer.Stderr(), "test")
+
+	container, err := appext.NewContainer(appContainer, "test", logger, verbosePrinter)
+	if err != nil {
+		return nil, err
+	}
+	omniProvider, err := bufmoduletesting.NewOmniProvider(
```

Comment:
If we're providing protovalidate via this `OmniProvider`, why do we still need to depend on it via a BSR dependency in `buftest`? Is there some functionality under the hood we are testing? It's not immediately apparent to me that, if the `*server` is using the `controller`, which itself is just depending on the `ModuleKeyProvider/ModuleDataProvider` given via `OmniProvider`, why we also need to have a hard dependency on `buf.build` here.

If not, we could get rid of `buftest` entirely, and move all the lsp testing code into `testdata` local to this package.

Reply:
OK, SGTM. Now `buftest` is no more, and I've reverted the changes made to the proto directory as well.